pkg/runtime/wfengine/backend.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package wfengine
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+
+	"github.com/microsoft/durabletask-go/api"
+	"github.com/microsoft/durabletask-go/backend"
+
+	"github.com/dapr/dapr/pkg/actors"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+)
+
+// workflowScheduler is an interface for pushing work items into the backend
+type workflowScheduler interface {
+	ScheduleWorkflow(wi *backend.OrchestrationWorkItem)
+	ScheduleActivity(wi *backend.ActivityWorkItem)
+}
+
+type actorBackend struct {
+	actors                    actors.Actors
+	orchestrationWorkItemChan chan *backend.OrchestrationWorkItem
+	activityWorkItemChan      chan *backend.ActivityWorkItem
+}
+
+func NewActorBackend() *actorBackend {
+	return &actorBackend{
+		orchestrationWorkItemChan: make(chan *backend.OrchestrationWorkItem),
+		activityWorkItemChan:      make(chan *backend.ActivityWorkItem),
+	}
+}
+
+func (be *actorBackend) SetActorRuntime(actors actors.Actors) {
+	be.actors = actors
+}
+
+// ScheduleActivity implements workflowScheduler
+func (be *actorBackend) ScheduleActivity(wi *backend.ActivityWorkItem) {
+	be.activityWorkItemChan <- wi
+}
+
+// ScheduleWorkflow implements workflowScheduler
+func (be *actorBackend) ScheduleWorkflow(wi *backend.OrchestrationWorkItem) {
+	be.orchestrationWorkItemChan <- wi
+}
+
+// CreateOrchestrationInstance implements backend.Backend and creates a new workflow instance.
+//
+// Internally, creating a workflow instance also creates a new actor with the same ID. The create
+// request is saved into the actor's "inbox" and then executed via a reminder thread. If the app is
+// scaled out across multiple replicas, the actor might get assigned to a replicas other than this one.
+func (be *actorBackend) CreateOrchestrationInstance(ctx context.Context, e *backend.HistoryEvent) error {
+	if err := be.validateConfiguration(); err != nil {
+		return err
+	}
+
+	workflowInstanceID := e.GetExecutionStarted().GetOrchestrationInstance().GetInstanceId()
```

Comment:
In looking at the durable task library, these are not guaranteed to be non-nil return values, which could cause this to catch on fire.

Reply:
You can't method chain and let a `try` block catch failures in Go -- trying to call a method on a nil reference will segfault so you will have to explicitly check (or better, return an error along with nil, which is arguably better if those things are error-ish)

Reply:
I need you to trust me that if any of these methods return `nil`, the result of the entire expression chain will be `""`. Every one of these methods test to see whether the receiver is `nil` and avoid segfaulting. When reading this generated code, it's very clear that the designers of protoc did this intentionally to allow this kind of chaining. I take advantage of this everywhere in the durabletask-go code and it works exactly as I expect, allowing the code to stay very concise and elegant.